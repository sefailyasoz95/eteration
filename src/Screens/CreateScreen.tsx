import { CommonActions } from "@react-navigation/routers";
import { NativeStackNavigationProp } from "@react-navigation/native-stack";
import React, { useState } from "react";
import { Keyboard, KeyboardAvoidingView, Platform, StyleSheet } from "react-native";
import Button from "../Components/Button/Button";
import Input from "../Components/Input/Input";
import { AppStackParamList } from "../StackParamLists/AppStackParamList";
import { formValidator } from "../Utils/Helpers/formValidator";
import { formErrorState } from "../Utils/InitialStates/formErrorState";
import { formValues } from "../Utils/InitialStates/formValues";
import { createSimpson } from "../Utils/Services/service";

type Props = {
	navigation: NativeStackNavigationProp<AppStackParamList, "Create">;
};

const CreateScreen = ({ navigation }: Props) => {
	const [loading, setLoading] = useState(false);
	const [formInputs, setFormInputs] = useState(formValues);
	const [errorStates, setErrorStates] = useState(formErrorState);
	const handleNewCharacter = async () => {
		setLoading(true);
		// burası kesinlikle daha iyi olabilir. Formik vb ekstra bir kütüphane kullanmak istemedim
		const { isformValid, tempErrors } = formValidator(formInputs);
		setErrorStates(tempErrors);
		if (isformValid) {
			let result = await createSimpson(formInputs);
			if (result.status === 200) {
				navigation.dispatch(
					CommonActions.reset({
						index: 1,
						routes: [{ name: "List" }],
					}),
				);
			}
		}
	};

	return (
		<KeyboardAvoidingView
			style={styles.container}
			onTouchStart={Keyboard.dismiss}
			behavior={Platform.OS === "ios" ? "padding" : "padding"}>
			<Input
				placeholder='Name Surname'
				testID='Name Surname'
				corner='rounded'
				onTextChanged={(value: string) => setFormInputs({ ...formInputs, name: value })}
				required={true}
				hasError={errorStates.nameSurname}
			/>
			<Input
				placeholder='Job Title'
				testID='Job Title'
				required={true}
				corner='rounded'
				hasError={errorStates.jobTitle}
				onTextChanged={(value: string) => setFormInputs({ ...formInputs, job: value })}
			/>
			<Input
				placeholder='About Him/Her'
				testID='About Him/Her'
				required={true}
				corner='rounded'
				type='textarea'
				hasError={errorStates.about}
				onTextChanged={(value: string) => setFormInputs({ ...formInputs, about: value })}
			/>
			<Input
				placeholder='Image Link'
				testID='Image Link'
				corner='rounded'
				onTextChanged={(value: string) => setFormInputs({ ...formInputs, avatar: value })}
			/>
			<Button
				textColor='#fff'
				text='Add Simpson'
				testID={"AddSimpson"}
				type='filled'
				onPress={handleNewCharacter}
				corner='curved'
				buttonColor='#000'
				textStyle={styles.buttonText}
				buttonStyle={styles.buttonContainer}
				loading={loading}
			/>
		</KeyboardAvoidingView>
	);
};

const styles = StyleSheet.create({
	container: {
		flex: 1,
		alignItems: "center",
		justifyContent: "flex-start",
		marginTop: 20,
	},
	buttonText: {
		fontWeight: "bold",
		fontSize: 16,
	},
	buttonContainer: {
		minWidth: 125,
	},
});

export default CreateScreen;
