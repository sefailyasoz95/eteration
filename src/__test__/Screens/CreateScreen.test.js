import React from "react";
import { act, fireEvent, render } from "@testing-library/react-native";
import CreateScreen from "../../Screens/CreateScreen";
import { shallow, configure } from "enzyme";
import toJson from "enzyme-to-json";
import { create } from "react-test-renderer";
import Adapter from "enzyme-adapter-react-16";
import { CommonActions } from "@react-navigation/routers";

configure({ adapter: new Adapter() });
const newSimpsonObject = {
	id: 0,
	name: "Sefa İlyas Öz",
	job: "Full Stack Developer",
	about:
		"Self motivated developer who tries to improve himself every day. Tries to be the best he can be at everything he does.",
	avatar:
		"https://media-exp1.licdn.com/dms/image/C4D03AQHexvLu9A-_gg/profile-displayphoto-shrink_800_800/0/1606719515595?e=1644451200&v=beta&t=nWK-Gy9V3oeP-i0-y-W4sEfqBEUU6zuubBHHjYEFiQo",
};
const navigation = {
	dispatch: jest.fn(),
};
describe("create screen", () => {
	it("should render screen correctly", () => {
		const component = shallow(<CreateScreen />);
		expect(component.length).toBe(1);
		expect(toJson(component)).toMatchSnapshot();
	});

	it("trying to sent an empty form with button press which will cause the inputs hasError prop to become true", () => {
		const component = create(<CreateScreen />);
		const button = component.root.findByProps({ testID: "AddSimpson" }).props;
		button.onPress();
		const nameSurnameInput = component.root.findByProps({ testID: "NameSurname" }).props;
		const jobTitle = component.root.findByProps({ testID: "JobTitle" }).props;
		const about = component.root.findByProps({ testID: "AboutHim/Her" }).props;
		const imageLink = component.root.findByProps({ testID: "ImageLink" }).props;

		nameSurnameInput.required
			? expect(nameSurnameInput.hasError).toBe(true)
			: expect(nameSurnameInput.hasError).toBe(false);

		jobTitle.required ? expect(jobTitle.hasError).toBe(true) : expect(jobTitle.hasError).toBe(false);

		about.required ? expect(about.hasError).toBe(true) : expect(about.hasError).toBe(false);

		imageLink.required ? expect(imageLink.hasError).toBe(true) : expect(imageLink.hasError).toBe(false || undefined);
	});

	it("creating new simpson", async () => {
		const component = create(<CreateScreen navigation={navigation} />);
		const button = component.root.findByProps({ testID: "AddSimpson" }).props;
		const nameSurnameInput = component.root.findByProps({ testID: "NameSurname" }).props;
		const jobTitle = component.root.findByProps({ testID: "JobTitle" }).props;
		const about = component.root.findByProps({ testID: "AboutHim/Her" }).props;
		await act(async () => {
			await nameSurnameInput.onTextChanged(newSimpsonObject.name);
		});
		await act(async () => {
			await jobTitle.onTextChanged(newSimpsonObject.job);
		});
		await act(async () => {
			await about.onTextChanged(newSimpsonObject.about);
		});
		await button.onPress();
		/* 
			Burası .not.toBeCalled() değil toBeCalled() olmalı tabii ki ancak CreateScreen.tsx içindeki handleNewCharacter
			çağırıldığında formInputs state'i initial state ile aynı geliyor ve if'e giremiyor, her ne kadar yukarıdaki 
			onTextChanged fonksiyonları çalışsa da... Onlar da şu şekilde çalışıyor, her biri state'deki objenin ilgili
			alanını set ediyor ancak "initialState"teki halini.
			CreateScreen.tsx'in içinde useEffect'i yorum satırı olarak bıraktım, yorum satırından kaldırır testi çalıştırırsanız 
			state'i set etme şeklini, anlatmaya çalıştığım şeyi daha iyi anlayabilirsiniz.
			bunu nasıl düzeltebiliriz, buna ne sebep oluyor, çözüm nedir gerçekten merak ediyorum. bir çözüm bulamadım.
			Bu süreç olumlu ya da olumsuz nasıl sonuçlanırsa sonuçlansın bu soruma cevap verebilirseniz çok sevinirim.
		*/
		expect(navigation.dispatch).not.toBeCalled();
	});
});
