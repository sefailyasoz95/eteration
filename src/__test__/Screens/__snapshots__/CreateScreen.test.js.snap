// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create screen should render screen correctly 1`] = `
<KeyboardAvoidingView
  behavior="padding"
  enabled={true}
  keyboardVerticalOffset={0}
  onTouchStart={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "flex-start",
      "marginTop": 20,
    }
  }
>
  <Input
    corner="rounded"
    hasError={false}
    onTextChanged={[Function]}
    placeholder="Name Surname"
    required={true}
    testID="NameSurname"
  />
  <Input
    corner="rounded"
    hasError={false}
    onTextChanged={[Function]}
    placeholder="Job Title"
    required={true}
    testID="JobTitle"
  />
  <Input
    corner="rounded"
    hasError={false}
    onTextChanged={[Function]}
    placeholder="About Him/Her"
    required={true}
    testID="AboutHim/Her"
    type="textarea"
  />
  <Input
    corner="rounded"
    onTextChanged={[Function]}
    placeholder="Image Link"
    testID="ImageLink"
  />
  <Button
    buttonColor="#000"
    buttonStyle={
      Object {
        "minWidth": 125,
      }
    }
    corner="curved"
    onPress={[Function]}
    testID="AddSimpson"
    text="Add Simpson"
    textColor="#fff"
    textStyle={
      Object {
        "fontSize": 16,
        "fontWeight": "bold",
      }
    }
    type="filled"
  />
</KeyboardAvoidingView>
`;
